mavros/src/plugins/hil.cpp

```diff
@@ -0,0 +1,320 @@
+				float rollspeed, float pitchspeed, float yawspeed,
+				int32_t lat, int32_t lon, int32_t alt,
+				int16_t vx, int16_t vy, int16_t vz,
+				uint16_t ind_airspeed, uint16_t true_airspeed,
+				int16_t xacc, int16_t yacc, int16_t zacc) {
+		set_hil_state_quaternion(stamp.toNSec() / 1000,
+					qw, qx, qy, qz,
+					rollspeed, pitchspeed, yawspeed,
+					lat, lon, alt,
+					vx, vy, vz,
+					ind_airspeed,
+					true_airspeed,
+					xacc, yacc, zacc);
+	}
+
+	/**
+	 * @brief Send hil_gps to FCU.
+	 */
+	void send_hil_gps(const ros::Time &stamp,
+				uint8_t fix_type,
+				int32_t lat, int32_t lon, int32_t alt,
+				uint16_t eph, uint16_t epv, uint16_t vel,
+				int16_t vn, int16_t ve, int16_t vd,
+				uint16_t cog,
+				uint8_t satellites_visible) {
+		set_hil_gps(stamp.toNSec() / 1000,
+					fix_type,
+					lat, lon, alt,
+					eph, epv, vel,
+					vn, ve, vd,
+					cog,
+					satellites_visible);
+	}
+
+	/**
+	 * @brief Send hil_sensor to FCU.
+	 */
+	void send_hil_sensor(const ros::Time &stamp,
+				float xacc, float yacc, float zacc,
+				float xgyro, float ygyro, float zgyro,
+				float xmag, float ymag, float zmag,
+				float abs_pressure, float diff_pressure, float pressure_alt,
+				float temperature,
+				uint32_t fields_updated) {
+		set_hil_sensor(stamp.toNSec() / 1000,
+					xacc, yacc, zacc,
+					xgyro, ygyro, zgyro,
+					xmag, ymag, zmag,
+					abs_pressure, diff_pressure, pressure_alt,
+					temperature,
+					fields_updated);
+	}
+
+	/* -*- callbacks -*- */
+
+	void state_quat_cb(const mavros_msgs::HilStateQuaternion::ConstPtr &req) {
+		send_hil_state_quaternion(req->header.stamp,
+					req->imu.orientation.w, req->imu.orientation.x, req->imu.orientation.y, req->imu.orientation.z,
+					req->imu.angular_velocity.x,req->imu.angular_velocity.y, req->imu.angular_velocity.z,
+					req->fix.latitude, req->fix.longitude, req->fix.altitude,
+					req->linear_velocity.x, req->linear_velocity.y, req->linear_velocity.z,
+					req->ind_airspeed,
+					req->true_airspeed,
+					req->imu.linear_acceleration.x, req->imu.linear_acceleration.y, req->imu.linear_acceleration.z);
+	}
+
+	void gps_cb(const mavros_msgs::HilGPS::ConstPtr &req) {
+		send_hil_gps(req->header.stamp,
+					req->fix_type,
+					req->fix.latitude, req->fix.longitude, req->fix.altitude,
+					req->eph, req->epv, req->vel,
+					req->vn, req->ve, req->vd,
+					req->cog,
+					req->satellites_visible);
+	}
+
+	void sensor_cb(const mavros_msgs::HilSensor::ConstPtr &req) {
+		// Throttle incoming messages to 40hz
+		if ((ros::Time::now() - last_time) < period) {
+			return;
+		}
+		last_time = ros::Time::now();
+
+		send_hil_sensor(req->header.stamp,
+					req->acc.x, req->acc.y, req->acc.z,
+					req->gyro.x, req->gyro.y, req->gyro.z,
+					req->mag.x, req->mag.y, req->mag.z,
+					req->abs_pressure.fluid_pressure, req->diff_pressure.fluid_pressure, req->pressure_alt.fluid_pressure,
+					req->temperature.temperature,
+					req->fields_updated);
```

Comment:
well having functions to send separate from the calls I think it is a good practice in terms of code perception. But ok I'll include the mavlink message fill on the cb.